File: src/orchestrator.js
/**
 * MASTER ORCHESTRATOR
 *
 * Purpose: Coordinate all tools in the hybrid team system
 * - Routes requests to appropriate tool
 * - Chains tool outputs (tool1 output → tool2 input)
 * - Maintains global state
 * - Manages cross-tool workflows
 *
 * Tools managed:
 * - map: Project + agent orchestration (19 agents, 5 phases — see src/server.js)
 * - [Other tools]: Extensible architecture
 *
 * Usage (local):
 * npm install
 * node src/orchestrator.js
 *
 * Usage (Vercel): deployed via api/orchestrator.js, mounted under the
 * /orchestrator prefix — see the prefix-stripping middleware below. MAP_URL
 * must point at the deployed MAP service's public URL (e.g. .../map) —
 * `http://localhost:3001` only works when both run locally.
 */

const express = require("express");
const cors = require("cors");
require("dotenv").config();

const { authenticateToken } = require("./middleware/auth");
const store = require("./store");

const app = express();
const PORT = process.env.ORCHESTRATOR_PORT || 3000;
const MAP_URL = process.env.MAP_URL || `http://localhost:${process.env.PORT || 3001}`;

// On Vercel this app is reached via a rewrite from /orchestrator/(.*) to
// /api/orchestrator, but the request URL the function sees is still the
// original /orchestrator/... path — strip the prefix so routes below can
// stay mounted at root either way.
app.use((req, res, next) => {
  if (req.url === "/orchestrator" || req.url.startsWith("/orchestrator/")) {
    req.url = req.url.slice(13) || "/";
  }
  next();
});

app.use(cors());
app.use(express.json());

// ============================================================================
// TOOL REGISTRY - Extend this as you add tools
// ============================================================================

const toolRegistry = {
  map: {
    name: "MAP",
    description: "Multi-Agent Project Manager — 19 agents across 5 SDLC phases",
    url: MAP_URL,
    available: ["projects", "agents", "phases", "orchestrate", "evaluate"],
    inputs: ["projectId", "phase", "step"],
    outputs: ["project", "agentResult", "evaluation"]
  }
};

// Health check stays unauthenticated — see src/server.js for the same rationale.
app.get("/health", (req, res) => {
  res.json({
    status: "ok",
    service: "orchestrator",
    timestamp: new Date().toISOString(),
    tools: Object.keys(toolRegistry)
  });
});

app.use(authenticateToken);

// ============================================================================
// REGISTRY
// ============================================================================

app.get("/tools", (req, res) => {
  res.json(Object.values(toolRegistry));
});

app.get("/tools/:name", (req, res) => {
  const tool = toolRegistry[req.params.name];
  if (!tool) return res.status(404).json({ error: "Tool not found" });
  res.json(tool);
});

// ============================================================================
// TOOL INVOCATION
// ============================================================================

async function callTool(toolName, action, body, token) {
  const tool = toolRegistry[toolName];
  if (!tool) {
    const err = new Error(`Tool not found: ${toolName}`);
    err.status = 404;
    throw err;
  }

  const response = await fetch(`${tool.url}/${action}`, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${token}`,
      "Content-Type": "application/json"
    },
    body: JSON.stringify(body)
  });

  if (!response.ok) {
    throw new Error(`Tool ${toolName} failed: ${response.statusText}`);
  }

  return response.json();
}

app.post("/tools/:toolName/:action", async (req, res) => {
  try {
    const { toolName, action } = req.params;
    const result = await callTool(toolName, action, req.body, req.token);
    await store.logToolInvocation({
      timestamp: new Date().toISOString(),
      tool: toolName,
      action,
      status: "success"
    });
    res.json({ tool: toolName, action, timestamp: new Date().toISOString(), result });
  } catch (error) {
    res.status(error.status || 500).json({ error: error.message });
  }
});

// ============================================================================
// TOOLCHAIN EXECUTION - Chain multiple tools
// ============================================================================

// Shared by /toolchain/execute AND /workflows/:id/execute — no internal HTTP
// call to itself (that relied on `localhost`, which serverless invocations
// don't share).
async function executeToolchain(sequence, data, token) {
  if (!sequence || !Array.isArray(sequence)) {
    const err = new Error("sequence must be array of {tool, action}");
    err.status = 400;
    throw err;
  }

  let currentData = data || {};
  const results = [];

  for (const step of sequence) {
    const { tool, action } = step;
    const result = await callTool(tool, action, currentData, token);
    results.push({ tool, action, result });

    currentData = step.passOutput ? result : currentData;
    await store.logToolInvocation({
      timestamp: new Date().toISOString(),
      tool,
      action,
      status: "success"
    });
  }

  return { sequence, results, finalData: currentData, timestamp: new Date().toISOString() };
}

app.post("/toolchain/execute", async (req, res) => {
  try {
    const { sequence, data } = req.body;
    const result = await executeToolchain(sequence, data, req.token);
    res.json(result);
  } catch (error) {
    res.status(error.status || 500).json({ error: error.message });
  }
});

// ============================================================================
// WORKFLOWS - Save and execute tool sequences
// ============================================================================

app.get("/workflows", async (req, res) => {
  try {
    res.json(await store.readWorkflows());
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

app.post("/workflows", async (req, res) => {
  try {
    const { name, description, sequence } = req.body;

    const workflow = {
      id: `workflow_${Date.now()}`,
      name,
      description,
      sequence,
      createdAt: new Date().toISOString()
    };

    const workflows = await store.readWorkflows();
    workflows.push(workflow);
    await store.writeWorkflows(workflows);

    res.status(201).json(workflow);
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

app.post("/workflows/:id/execute", async (req, res) => {
  try {
    const workflows = await store.readWorkflows();
    const workflow = workflows.find((w) => w.id === req.params.id);

    if (!workflow) {
      return res.status(404).json({ error: "Workflow not found" });
    }

    const execution = await executeToolchain(workflow.sequence, req.body.data || {}, req.token);
    res.json({ workflow: workflow.name, execution });
  } catch (error) {
    res.status(error.status || 500).json({ error: error.message });
  }
});

// ============================================================================
// SYSTEM STATE
// ============================================================================

app.get("/system/state", (req, res) => {
  res.json({
    tools: Object.keys(toolRegistry),
    port: PORT,
    storage: store.usingKV ? "vercel-kv" : "filesystem",
    timestamp: new Date().toISOString()
  });
});

// ============================================================================
// START (local/self-hosted only — Vercel requires the app, not a listener)
// ============================================================================

if (require.main === module) {
  app.listen(PORT, () => {
    console.log(`Master Orchestrator running on port ${PORT}`);
    console.log(`Tools: ${Object.keys(toolRegistry).join(", ")}`);
    console.log(`MAP_URL: ${MAP_URL}`);
  });
}

module.exports = app;

File: app/routers/orchestrator.py
"""
MASTER ORCHESTRATOR — migrated from src/orchestrator.js.

Note on auth: unlike agents.py/projects.py (which require the auth
dependency), these routes are left unauthenticated here, matching how this
router is being wired up today (no `Depends(authenticate_token)` on the
APIRouter). The JS original did gate everything except /health behind
authenticateToken and forwarded req.token to callTool/executeToolchain — if
auth is added back later, thread the validated token through the same way
agents.py does.
"""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Any

from fastapi import APIRouter, Body, HTTPException, Request

from app.core.config import get_settings
from app.core.storage import get_storage
from app.schemas.orchestrator import (
    ToolchainRequest,
    Workflow,
    WorkflowCreate,
    WorkflowExecuteRequest,
)
from app.services.orchestrator import tool_registry, workflows_store
from app.services.orchestrator.toolchain import execute_toolchain
from app.services.orchestrator.tool_registry import ToolError

router = APIRouter()


def _now() -> str:
    return datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")


def _token_from_request(request: Request) -> str:
    """No auth dependency is mounted on this router (see module docstring),
    so there's no validated token from a security scheme — pass through
    whatever bearer token (if any) the caller sent, same value the JS
    original forwarded to callTool via req.token."""
    auth_header = request.headers.get("authorization", "")
    if auth_header.lower().startswith("bearer "):
        return auth_header[7:]
    return ""


# ============================================================================
# HEALTH — stays unauthenticated, same rationale as app/routers/health.py
# ============================================================================


@router.get("/health")
async def health() -> dict[str, Any]:
    return {
        "status": "ok",
        "service": "orchestrator",
        "timestamp": _now(),
        "tools": list(tool_registry.get_tool_registry().keys()),
    }


# ============================================================================
# REGISTRY
# ============================================================================


@router.get("/tools")
async def list_tools() -> list[dict[str, Any]]:
    return [tool.model_dump() for tool in tool_registry.list_tools()]


@router.get("/tools/{name}")
async def get_tool(name: str) -> dict[str, Any]:
    tool = tool_registry.get_tool(name)
    if tool is None:
        raise HTTPException(status_code=404, detail="Tool not found")
    return tool.model_dump()


# ============================================================================
# TOOL INVOCATION
# ============================================================================


@router.post("/tools/{tool_name}/{action}")
async def invoke_tool(
    tool_name: str,
    action: str,
    request: Request,
    body: dict[str, Any] = Body(default={}),
) -> dict[str, Any]:
    token = _token_from_request(request)
    try:
        result = await tool_registry.call_tool(tool_name, action, body, token)
    except ToolError as error:
        raise HTTPException(status_code=error.status, detail=str(error)) from error

    get_storage().log_tool_invocation(
        {
            "timestamp": _now(),
            "tool": tool_name,
            "action": action,
            "status": "success",
        }
    )
    return {"tool": tool_name, "action": action, "timestamp": _now(), "result": result}


# ============================================================================
# TOOLCHAIN EXECUTION — chain multiple tools
# ============================================================================


@router.post("/toolchain/execute")
async def toolchain_execute(request: Request, body: ToolchainRequest) -> dict[str, Any]:
    token = _token_from_request(request)
    try:
        return await execute_toolchain(body.sequence, body.data, token)
    except ToolError as error:
        raise HTTPException(status_code=error.status, detail=str(error)) from error


# ============================================================================
# WORKFLOWS — save and execute tool sequences
# ============================================================================


@router.get("/workflows")
async def list_workflows() -> list[dict[str, Any]]:
    return workflows_store.list_workflows()


@router.post("/workflows", status_code=201)
async def create_workflow(payload: WorkflowCreate) -> dict[str, Any]:
    return workflows_store.create_workflow(payload)


@router.post("/workflows/{workflow_id}/execute")
async def execute_workflow(
    workflow_id: str, request: Request, body: WorkflowExecuteRequest
) -> dict[str, Any]:
    workflow = workflows_store.get_workflow(workflow_id)
    if workflow is None:
        raise HTTPException(status_code=404, detail="Workflow not found")

    token = _token_from_request(request)
    sequence = Workflow.model_validate(workflow).sequence
    try:
        execution = await execute_toolchain(sequence, body.data, token)
    except ToolError as error:
        raise HTTPException(status_code=error.status, detail=str(error)) from error

    return {"workflow": workflow.get("name"), "execution": execution}


# ============================================================================
# SYSTEM STATE
# ============================================================================


@router.get("/system/state")
async def system_state() -> dict[str, Any]:
    settings = get_settings()
    storage = get_storage()
    return {
        "tools": list(tool_registry.get_tool_registry().keys()),
        "port": settings.ORCHESTRATOR_PORT,
        "storage": "vercel-kv" if storage.using_kv else "filesystem",
        "timestamp": _now(),
    }
